docs/narr/advfeatures.rst

```diff
@@ -0,0 +1,399 @@
+to templates, request objects, and other niceties.
+
+The Pyramid debug toolbar is a "tween", as is the ``pyramid_tm`` transaction
+manager.
+
+Example: :ref:`registering_tweens`.
+
+Return What You Want From Your Views
+------------------------------------
+
+We have shown before (in the :doc:`introduction`) how using a :term:`renderer`
+allows you to return simple Python dictionaries from your view code. But some
+frameworks allow you to return strings or tuples from view callables.
+When frameworks allow for this, code looks slightly prettier, because there are
+fewer imports, and less code. For example, compare this:
+
+.. code-block:: python
+   :linenos:
+
+   from pyramid.response import Response
+
+   def aview(request):
+       return Response("Hello world!")
+
+To this:
+
+.. code-block:: python
+   :linenos:
+
+   def aview(request):
+       return "Hello world!"
+
+Nicer to look at, right?
+
+Out of the box, Pyramid will raise an exception if you try to run the second
+example above. After all, a view should return a response, and "explicit is
+better than implicit".
+
+But if you're a developer who likes the aesthetics of simplicity, Pyramid
+provides an way to support this sort of thing, the *response adapter*:
+
+.. code-block:: python
+   :linenos:
+
+   from pyramid.config import Configurator
+   from pyramid.response import Response
+
+   def string_response_adapter(s):
+       response = Response(s)
+       response.content_type = 'text/html'
+       return response
+
+A new response adapter is registered in configuration:
+
+   if __name__ == '__main__':
+       config = Configurator()
+       config.add_response_adapter(string_response_adapter, basestring)
+
+With that, you may return strings from any of your view callables, e.g.:
+
+.. code-block:: python
+   :linenos:
+
+   def helloview(request):
+       return "Hello world!"
+
+   def goodbyeview(request):
+       return "Goodbye world!"
+
+You can even use a response adapter to allow for custom content types and
+return codes:
+
+.. code-block:: python
+   :linenos:
+
+   from pyramid.config import Configurator
+
+   def tuple_response_adapter(val):
+       status_int, content_type, body = val
+       response = Response(body)
+       response.content_type = content_type
+       response.status_int = status_int
+       return response
+
+   def string_response_adapter(body):
+       response = Response(body)
+       response.content_type = 'text/html'
+       response.status_int = 200
+       return response
+
+   if __name__ == '__main__':
+       config = Configurator()
+       config.add_response_adapter(string_response_adapter, basestring)
+       config.add_response_adapter(tuple_response_adapter, tuple)
+
+With this, both of these views will work as expected:
+
+.. code-block:: python
+   :linenos:
+
+   def aview(request):
+       return "Hello world!"
+
+   def anotherview(request):
+       return (403, 'text/plain', "Forbidden")
+
```

Comment:
Again, this is because the three-space indent is part of the `.. code-block:: directive`.